plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/domains/UnifiedTypeDomain.kt

```diff
@@ -0,0 +1,444 @@
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.domains
+
+import org.jetbrains.kotlin.formver.embeddings.ClassTypeEmbedding
+import org.jetbrains.kotlin.formver.viper.MangledName
+import org.jetbrains.kotlin.formver.viper.ast.*
+
+
+const val UNIFIED_TYPE_DOMAIN_NAME = "UnifiedType"
+
+
+/**
+ * This new domain is designed to replace `NullableDomain`, `TypeDomain` and `CastingDomain` and it is not yet integrated.
+ * To enable its generation in viper output uncomment corresponding lines in
+ * [ProgramConverter](jetbrains://idea/navigate/reference?project=kotlin&path=org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:70)
+ * and [SpecialFunctions.kt](jetbrains://idea/navigate/reference?project=kotlin&path=org/jetbrains/kotlin/formver/embeddings/callables/SpecialFunctions.kt:58)
+ *
+ * Viper code:
+ * ```viper
+ *
+ * domain UnifiedType  {
+ *
+ *
+ *  unique function intType(): UnifiedType
+ *  unique function boolType(): UnifiedType
+ *  unique function unitType(): UnifiedType
+ *  unique function nothingType(): UnifiedType
+ *  unique function anyType(): UnifiedType
+ *  unique function functionType(): UnifiedType
+ *
+ *  // unique *Type() : UnifiedType for each user type
+ *
+ *  function nullValue(): Ref
+ *  function unitValue(): Ref
+ *
+ *  function isSubtype(t1: UnifiedType, t2: UnifiedType): Bool
+ *  function typeOf(r: Ref): UnifiedType
+ *  function nullable(t: UnifiedType): UnifiedType
+ *
+ *
+ *  function intToRef(v: Int): Ref
+ *  function intFromRef(r: Ref): Int
+ *  function boolToRef(v: Bool): Ref
+ *  function boolFromRef(r: Ref): Bool
+ *
+ *
+ *  axiom subtype_reflexive {
+ *    (forall t: UnifiedType ::isSubtype(t, t))
+ *  }
+ *
+ *  axiom subtype_transitive {
+ *    (forall t1: UnifiedType, t2: UnifiedType, t3: UnifiedType ::
+ *      { isSubtype(t1, t2), isSubtype(t2, t3) }
+ *      isSubtype(t1, t2) &&
+ *      isSubtype(t2, t3) ==>
+ *      isSubtype(t1, t3))
+ *  }
+ *
+ *  axiom subtype_antisymmetric {
+ *    (forall t1: UnifiedType, t2: UnifiedType ::
+ *      { isSubtype(t1, t2), isSubtype(t2, t1) }
+ *      isSubtype(t1, t2) &&
+ *      isSubtype(t2, t1) ==>
+ *      t1 == t2)
+ *  }
+ *
+ *  axiom nullable_idempotent {
+ *    (forall t: UnifiedType ::
+ *      { nullable(nullable(t)) }
+ *      nullable(nullable(t)) ==
+ *      nullable(t))
+ *  }
+ *
+ *  axiom nullable_supertype {
+ *    (forall t: UnifiedType ::
+ *      { nullable(t) }
+ *      isSubtype(t, nullable(t)))
+ *  }
+ *
+ *  axiom nullable_preserves_subtype {
+ *    (forall t1: UnifiedType, t2: UnifiedType ::
+ *      { isSubtype(nullable(t1), nullable(t2)) }
+ *      isSubtype(t1, t2) ==>
+ *      isSubtype(nullable(t1), nullable(t2)))
+ *  }
+ *
+ *  axiom nullable_any_supertype {
+ *    (forall t: UnifiedType ::isSubtype(t, nullable(anyType())))
+ *  }
```

Comment:
I wonder whether this is necessary, since we know `t <: Any` and `a <: b` implies `nullable(a) <: nullable(b)`.  At some point it would be good to check the performance.

Reply:
Since there is no axiom that each `t` is either subtype of `Any` or nullable version of such a type, I'm not sure compiler can deduce it for arbitrary type. On other hand, it seems that we never have arbitrary types.

Reply:
That's a good point.  Probably good to keep this one, then until we have a better understanding of generics.